File: app/src/main/java/com/kelsos/mbrc/RemoteApplication.java
package com.kelsos.mbrc;

import android.app.Application;
import android.support.annotation.CallSuper;

import com.kelsos.mbrc.di.modules.RemoteModule;

import timber.log.Timber;
import toothpick.Scope;
import toothpick.Toothpick;
import toothpick.configuration.Configuration;
import toothpick.registries.FactoryRegistryLocator;
import toothpick.registries.MemberInjectorRegistryLocator;
import toothpick.smoothie.module.SmoothieApplicationModule;
import uk.co.chrisjenx.calligraphy.CalligraphyConfig;

public class RemoteApplication extends Application {

  @CallSuper
  public void onCreate() {
    super.onCreate();

    Configuration configuration;
    if (BuildConfig.DEBUG) {
      configuration = Configuration.forDevelopment().disableReflection();
    } else {
      configuration = Configuration.forProduction().disableReflection();
    }

    Toothpick.setConfiguration(configuration);

    MemberInjectorRegistryLocator.setRootRegistry(new MemberInjectorRegistry());
    FactoryRegistryLocator.setRootRegistry(new FactoryRegistry());
    Scope applicationScope = Toothpick.openScope(this);
    applicationScope.installModules(new SmoothieApplicationModule(this), new RemoteModule());

    CalligraphyConfig.initDefault(new CalligraphyConfig.Builder()
        .setDefaultFontPath("fonts/roboto_regular.ttf")
        .setFontAttrId(R.attr.fontPath)
        .build());

    if (BuildConfig.DEBUG) {
      Timber.plant(new Timber.DebugTree() {
        @Override protected String createStackElementTag(StackTraceElement element) {
          return super.createStackElementTag(element) + ":" +
              element.getLineNumber() +
              " [" + Thread.currentThread().getName() + "]";
        }
      });
    }

  }
}

File: app/src/main/kotlin/com/kelsos/mbrc/RemoteApplication.kt
package com.kelsos.mbrc


import androidx.annotation.CallSuper
import androidx.multidex.MultiDexApplication
import com.jakewharton.threetenabp.AndroidThreeTen
import com.kelsos.mbrc.di.modules.RemoteModule
import com.raizlabs.android.dbflow.config.FlowConfig
import com.raizlabs.android.dbflow.config.FlowManager
import timber.log.Timber
import toothpick.Toothpick
import toothpick.configuration.Configuration
import toothpick.registries.FactoryRegistryLocator
import toothpick.registries.MemberInjectorRegistryLocator
import toothpick.smoothie.module.SmoothieApplicationModule

open class RemoteApplication : MultiDexApplication() {

  @CallSuper
  override fun onCreate() {
    super.onCreate()
    initialize()
  }

  protected open fun initialize() {
    if (!testMode()) {
      AndroidThreeTen.init(this)
    }
    initializeDbflow()
    initializeToothpick()
    initializeTimber()
  }


  private fun initializeTimber() {
    if (BuildConfig.DEBUG) {
      Timber.plant(object : Timber.DebugTree() {
        override fun createStackElementTag(element: StackTraceElement): String {
          return "${super.createStackElementTag(element)}:${element.lineNumber} [${Thread.currentThread().name}]"
        }
      })
    }
  }

  private fun initializeDbflow() {
    FlowManager.init(FlowConfig.Builder(this).openDatabasesOnInit(true).build())
  }

  private fun initializeToothpick() {
    val configuration: Configuration = if (BuildConfig.DEBUG) {
      Configuration.forDevelopment().disableReflection()
    } else {
      Configuration.forProduction().disableReflection()
    }

    Toothpick.setConfiguration(configuration)

    MemberInjectorRegistryLocator.setRootRegistry(MemberInjectorRegistry())
    FactoryRegistryLocator.setRootRegistry(FactoryRegistry())
    val applicationScope = Toothpick.openScope(this)
    if (!testMode()) {
      applicationScope.installModules(SmoothieApplicationModule(this), RemoteModule())
    }
  }


  internal open fun testMode(): Boolean = false
}
